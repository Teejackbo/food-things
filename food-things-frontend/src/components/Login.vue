<template>
  <Spinner v-if="submitting" :fullScreen="true" />
  <div v-else>
    <h1>Login</h1>
    <form @submit.prevent="handleSubmit">
      <InputField
        v-model="userData.username"
        type="text"
        placeholder="joe_bloggs"
        label="Username"
        icon="user"
      />

      <InputField
        v-model="userData.password"
        type="password"
        placeholder="•••••••••"
        label="Password"
        icon="key"
      />

      <InputField
        v-model="remember"
        type="checkbox"
        label="Remember Me"
      />

      <PrimaryButton type="success">Login</PrimaryButton>
    </form>
  </div>
</template>

<script>
import { mapActions, mapState } from 'vuex'
import { LOGIN_ACTION } from '@/store/modules/user/types'

export default {
  name: 'Login',

  data: () => ({
    userData: {
      username: '',
      password: '',
    },
    submitting: false,
    remember: false,
  }),

  computed: mapState({
    user: state => state.user.user,
  }),

  methods: {
    ...mapActions({
      login: LOGIN_ACTION,
    }),

    async handleSubmit() {
      this.submitting = true
      await this.login({
        loginInfo: this.userData,
        remember: this.remember,
        alert: true,
      })
      this.submitting = false

      if (this.user) this.$router.push({ name: 'home' })
    },
  },
}
</script>
