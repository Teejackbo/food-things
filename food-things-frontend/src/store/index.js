import Vue from 'vue'
import Vuex from 'vuex'

import user from '@/store/modules/user'
import alerts from '@/store/modules/alerts'
import food from '@/store/modules/food'

Vue.use(Vuex)

export default new Vuex.Store({
  modules: {
    user,
    alerts,
    food,
  },
  state: {},
  mutations: {},
  actions: {},
})
