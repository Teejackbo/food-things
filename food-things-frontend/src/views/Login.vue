<template>
  <Spinner v-if="submitting" :fullScreen="true" />
  <div v-else>
    <h1>Login</h1>
    <form @submit.prevent="handleSubmit">
      <InputField
        v-model="userData.username"
        type="text"
        placeholder="joe_bloggs"
        label="Username"
        name="username"
        icon="user"
      />

      <InputField
        v-model="userData.password"
        type="password"
        placeholder="•••••••••"
        label="Password"
        name="password"
        icon="key"
      />

      
      <label>
        Remember Me?
        <Checkbox v-model="remember" />
      </label>

      <PrimaryButton type="success" data-cy-login-btn>Login</PrimaryButton>
    </form>
  </div>
</template>

<script>
import { mapActions, mapState } from 'vuex'
import { ADD_ALERT_ACTION } from '@/store/modules/alerts/types'
import { LOGIN_ACTION } from '@/store/modules/user/types'

export default {
  name: 'LoginView',

  data: () => ({
    userData: {
      username: '',
      password: '',
    },
    submitting: false,
    remember: false,
  }),

  computed: mapState({
    id: state => state.user.user.id,
  }),

  created() {
    if (this.id) {
      this.$router.replace({ name: 'home' })
      this.alert({
        message: 'You are already logged in.',
        type: 'error',
        timeout: 3000,
      })
    }
  },

  methods: {
    ...mapActions({
      alert: ADD_ALERT_ACTION,
      login: LOGIN_ACTION,
    }),

    async handleSubmit() {
      this.submitting = true
      await this.login({
        loginInfo: this.userData,
        remember: this.remember,
        alert: true,
      })
      this.submitting = false

      if (this.id) this.$router.push({ name: 'home' })
    },
  },
}
</script>
