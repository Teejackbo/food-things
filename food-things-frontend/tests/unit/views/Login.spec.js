import { shallowMount } from '@vue/test-utils'
import Login from '@/views/Login'
import registerGlobals from '../components/global'
import $store from '@/store'

registerGlobals()

describe('<Login />', () => {
  it('Should match snapshot.', () => {
    const wrapper = shallowMount(Login, { mocks: { $store } })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('Should display a fullscreen spinner if it is submitting.', () => {
    const wrapper = shallowMount(Login, {
      data() {
        return {
          submitting: true,
        }
      },
      mocks: {
        $store,
      },
    })

    expect(wrapper.attributes()).toEqual({ fullscreen: 'true' })
    expect(wrapper.element).toMatchSnapshot()
  })
})
