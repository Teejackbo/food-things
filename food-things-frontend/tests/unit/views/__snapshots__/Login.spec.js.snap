// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> Should display a fullscreen spinner if it is submitting. 1`] = `
<vuecomponent-stub
  fullscreen="true"
/>
`;

exports[`<Login /> Should match snapshot. 1`] = `
<div>
  <h1>
    Login
  </h1>
   
  <form>
    <vuecomponent-stub
      name="username"
      placeholder="joe_bloggs"
    />
     
    <validationmessage-stub
      style="display: none;"
    />
     
    <vuecomponent-stub
      name="password"
      placeholder="•••••••••"
      type="password"
    />
     
    <validationmessage-stub
      style="display: none;"
    />
     
    <label>
      
      Remember Me?
      
      <vuecomponent-stub />
    </label>
     
    <vuecomponent-stub
      data-cy-login-btn=""
      type="success"
    />
  </form>
</div>
`;
